File: mau_ref_global.py
"""
Replica of mau_ref_global (R): total global MAU Prophet forecast with DperMar regressor.

Dependencies:
  pip3 install pandas prophet matplotlib openpyxl
  /Users/grac/Documents/GitHub/profit-from-prophet/.venv/bin/python -m pip install matplotlib
  /Users/grac/Documents/GitHub/profit-from-prophet/.venv/bin/python -m pip install pandas
  /Users/grac/Documents/GitHub/profit-from-prophet/.venv/bin/python -m pip install prophet
  /Users/grac/Documents/GitHub/profit-from-prophet/.venv/bin/python -m pip install openpyxl

Adjust IN_DIR, OUT_DIR, and input filenames to match your Box paths.
"""
from __future__ import annotations

import calendar
import os
from datetime import date, timedelta

import matplotlib.pyplot as plt
import pandas as pd
from prophet import Prophet

# ---------------------------------------------------------------------------
# Dates (same as R)
# ---------------------------------------------------------------------------
D_ACT_START = date(2022, 1, 1)
D_ACT_END = date(2026, 1, 31)
D_FSTART = D_ACT_END + timedelta(days=1)
D_FEND = D_FSTART + timedelta(days=364 * 3)  # R: D_FEnd <- D_FStart + 364*3
# R: Scenario <- "202601" (reference only)
D_TODAY = date.today().strftime("%Y.%m.%d")

# ---------------------------------------------------------------------------
# Directories — edit to match your machine (R used Box paths)
# ---------------------------------------------------------------------------
IN_DIR = os.path.expanduser(
    "/Users/grac/Library/CloudStorage/Box-Box/Internal Reporting (CPL)/FP&A/"
    "Modelling/03. MAU/07. Prophet Model/02. Prophet Inputs/"
)
OUT_DIR = os.path.expanduser(
    "/Users/grac/Library/CloudStorage/Box-Box/Internal Reporting (CPL)/FP&A/"
    "Modelling/03. MAU/07. Prophet Model/03. Prophet Outputs/"
)

# ---------------------------------------------------------------------------
# Metric selection
# ---------------------------------------------------------------------------
METRIC = "MAU"  # or "Signups"
COL = "TOTAL_SIGNUPS" if METRIC == "Signups" else "TOTAL_MAU"
FILE_MAIN = "2026.02.16 Signups.csv" if METRIC == "Signups" else "2026.02.20 MAU.csv"


def days_in_month(dt: pd.Timestamp) -> int:
    return calendar.monthrange(dt.year, dt.month)[1]


def main() -> None:
    os.makedirs(OUT_DIR, exist_ok=True)

    path_main = os.path.join(IN_DIR, FILE_MAIN)
    path_ms_a = os.path.join(IN_DIR, "2026.02.16 Looker_MS_A_D.csv")
    path_ms_f = os.path.join(IN_DIR, "2026.02.16 MS_F_M.csv")
    path_hol = os.path.join(IN_DIR, "2026.01.05 collated_holidays.csv")

    df = pd.read_csv(path_main)
    ms_a_d = pd.read_csv(path_ms_a)
    ms_f_m = pd.read_csv(path_ms_f)
    hol_raw = pd.read_csv(path_hol)

    # Actuals
    a = (
        df.assign(ds=pd.to_datetime(df["DATE"], format="%m/%d/%Y").dt.normalize())
        .loc[lambda x: (x["ds"].dt.date >= D_ACT_START) & (x["ds"].dt.date <= D_ACT_END)]
        .groupby("ds", as_index=False)[COL]
        .sum(min_count=1)
        .rename(columns={COL: "y"})
    )
    full_act = pd.DataFrame(
        {"ds": pd.date_range(D_ACT_START, D_ACT_END, freq="D")}
    )
    a = full_act.merge(a, on="ds", how="left")
    a["y"] = a["y"].fillna(0)

    # Regressor: actual daily marketing
    ds_ms_a_d = (
        ms_a_d.assign(ds=pd.to_datetime(ms_a_d["ds"], format="%m/%d/%Y").dt.normalize())
        .loc[lambda x: x["ds"].dt.date <= D_ACT_END]
        .groupby("ds", as_index=False)["perMar"]
        .sum(min_count=1)
        .rename(columns={"perMar": "DperMar"})
    )

    # Monthly forecast budget → daily spread in forecast window
    ms_f_m = ms_f_m.copy()
    ms_f_m["MONTH"] = pd.to_datetime(ms_f_m["MONTH"], format="%m/%d/%Y")
    ds_ms_f_m = (
        ms_f_m.loc[ms_f_m["BUDGET"] == "Performance Marketing"]
        .assign(m=lambda x: x["MONTH"].dt.to_period("M").dt.to_timestamp())
        .groupby("m", as_index=False)["TARGET"]
        .sum(min_count=1)
        .rename(columns={"TARGET": "pm"})
    )

    t_full = pd.DataFrame({"ds": pd.date_range(D_ACT_START, D_FEND, freq="D")})

    t_fstart = pd.DataFrame({"ds": pd.date_range(D_FSTART, D_FEND, freq="D")})
    t_fstart["m"] = t_fstart["ds"].dt.to_period("M").dt.to_timestamp()
    forecast_daily = t_fstart.merge(ds_ms_f_m, on="m", how="left")
    forecast_daily["DperMar"] = (
        forecast_daily["pm"].fillna(0)
        / forecast_daily["m"].map(lambda d: days_in_month(pd.Timestamp(d)))
    )
    forecast_daily = forecast_daily[["ds", "DperMar"]]

    r = t_full.merge(ds_ms_a_d, on="ds", how="left")
    r = r.merge(forecast_daily, on="ds", how="left", suffixes=("", "_y"))
    r["DperMar"] = r["DperMar"].combine_first(r["DperMar_y"]).fillna(0)
    r = r[["ds", "DperMar"]]

    # Training data
    train = (
        r.merge(a, on="ds", how="left")
        .loc[lambda x: x["ds"].dt.date <= D_ACT_END][["ds", "y", "DperMar"]]
        .assign(
            ds=lambda x: pd.to_datetime(x["ds"]),
            y=lambda x: pd.to_numeric(x["y"], errors="coerce"),
            DperMar=lambda x: pd.to_numeric(
                x["DperMar"].astype(str).str.replace(",", "", regex=False),
                errors="coerce",
            ),
        )
        .dropna(subset=["ds"])
        .sort_values("ds")
        .reset_index(drop=True)
    )

    # US holidays for Prophet
    us_hol = hol_raw.loc[hol_raw["country"] == "US"].copy()
    holidays_df = us_hol.assign(
        ds=pd.to_datetime(us_hol["ds"], format="%m/%d/%Y")
    )[["holiday", "ds"]].reset_index(drop=True)

    prophet_kw = dict(
        changepoint_range=0.7,
        daily_seasonality=False,
        weekly_seasonality=4,
        yearly_seasonality=10,
        changepoint_prior_scale=0.05,
        seasonality_prior_scale=10,
    )
    if not holidays_df.empty:
        prophet_kw["holidays"] = holidays_df
    m = Prophet(**prophet_kw)
    m.add_regressor("DperMar", standardize=False)
    m.fit(train)

    fc = m.predict(r)

    f = fc[["ds", "yhat"]].rename(columns={"yhat": "MAU"})
    f["ds"] = pd.to_datetime(f["ds"]).dt.normalize()

    monthly = (
        f.assign(Date=lambda x: x["ds"] + pd.offsets.MonthEnd(0))
        .groupby("Date", as_index=False)["MAU"]
        .sum()
    )

    out_xlsx = os.path.join(OUT_DIR, f"{D_TODAY}_Outputs_Total_Prophet_DperMar.xlsx")
    with pd.ExcelWriter(out_xlsx, engine="openpyxl") as writer:
        f.to_excel(writer, sheet_name="Total_Daily", index=False)
        monthly.to_excel(writer, sheet_name="Total_Monthly", index=False)
    print(f"Wrote {out_xlsx}")

    fig1 = m.plot(fc)
    fig1.suptitle(f"Forecast for Total {METRIC}")
    fig1.axes[0].set_xlabel("Year")
    fig1.axes[0].set_ylabel("Total")
    plt.tight_layout()
    plt.show()

    fig2 = m.plot_components(fc)
    fig2.suptitle(f"Prophet Components: {METRIC}")
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    main()

File: mau_ref_global.r
#Installing Packages

#Tables
install.packages('data.table') #Enhanced version of data frame
install.packages('tibble') #Simple data frames

#Cleaning & Transformation
install.packages('lubridate') #Date and time transformations
install.packages('magrittr') #%>% 
install.packages('stringr') #String mutations
#install.packages('plyr') 
install.packages('dplyr') 
install.packages('tidyr') 
install.packages('tidyverse') 
install.packages('mgsub') #Multiple Text Substitution 
install.packages("DatabaseConnector") #eoMonth function

#Modelling
install.packages('prophet')

#Charting
install.packages('ggplot2')

#Exporting
install.packages('openxlsx')

#Running Packages
#library(plyr)
library(data.table)
library(lubridate)
library(magrittr)
library(dplyr)
library(stringr)
library(tibble)
library(tidyr)
library(tidyverse)
library(ggplot2)
library(prophet)
library(mgsub)
library(DatabaseConnector) #why is this not loading
library(openxlsx)

# Dates
D_ActStart <- as.Date("2022-01-01")
D_ActEnd <- as.Date("2026-01-31")
D_FStart <- D_ActEnd+1
D_FEnd <- D_FStart+364*3
Scenario <- "202601"
D_Today <- format(Sys.Date(),"%Y.%m.%d")

# Directories
in_dir <- "/Users/grac/Library/CloudStorage/Box-Box/Internal Reporting (CPL)/FP&A/Modelling/03. MAU/07. Prophet Model/02. Prophet Inputs/"
out_dir <- "/Users/grac/Library/CloudStorage/Box-Box/Internal Reporting (CPL)/FP&A/Modelling/03. MAU/07. Prophet Model/03. Prophet Outputs/"

# Metric Selection
#metric <- "Signups"
metric <- "MAU"
col  <- if (metric == "Signups") "TOTAL_SIGNUPS" else "TOTAL_MAU"

# Input Files
file <- if (metric == "Signups") "2026.02.16 Signups.csv" else "2026.02.20 MAU.csv"
DF <- as_tibble(fread(file.path(in_dir, file)))
MS_A_D <- as_tibble(fread(paste0(in_dir,"2026.02.16 Looker_MS_A_D.csv")))
MS_F_M <- as_tibble(fread(paste0(in_dir,"2026.02.16 MS_F_M.csv")))
hol_raw <- as_tibble(fread(paste0(in_dir,"2026.01.05 collated_holidays.csv")))

# Actuals 
A <- DF %>%
  mutate(ds = as.Date(DATE, "%m/%d/%Y")) %>%
  filter(ds >= D_ActStart, ds <= D_ActEnd) %>%
  group_by(ds) %>%
  summarise(y = sum(.data[[col]], na.rm = TRUE), .groups = "drop") %>%
  complete(ds = seq.Date(D_ActStart, D_ActEnd, by = "day")) %>%
  mutate(y = replace_na(y, 0))

# Regressor Table
DS_MS_A_D <- MS_A_D %>%
  mutate(ds = as.Date(ds, "%m/%d/%Y")) %>%
  filter(ds <= D_ActEnd) %>%
  group_by(ds) %>% summarise(DperMar = sum(perMar, na.rm = TRUE), .groups = "drop")

DS_MS_F_M <- MS_F_M %>%
  mutate(MONTH = as.Date(MONTH, "%m/%d/%Y")) %>%
  filter(BUDGET == "Performance Marketing") %>%
  mutate(m = floor_date(MONTH, "month")) %>%
  group_by(m) %>% summarise(pm = sum(TARGET, na.rm = TRUE), .groups = "drop")

T_Full <- tibble(ds = seq.Date(D_ActStart, D_FEnd, "day")) 

T_FStart <- tibble(ds = seq.Date(D_FStart, D_FEnd, "day")) %>%
  mutate(m = floor_date(ds, "month"))

R <- T_Full %>%
  left_join(DS_MS_A_D, by = "ds") %>%
  left_join(T_FStart %>%
      left_join(DS_MS_F_M, by = "m") %>%
      transmute(ds, DperMar = replace_na(pm, 0) / days_in_month(m)), by = "ds") %>%
  transmute(ds, DperMar = coalesce(DperMar.x, DperMar.y, 0))

# Training Data
train <- R %>%
  left_join(A, by = "ds") %>%
  filter(ds <= D_ActEnd) %>%
  select(ds, y, DperMar) %>%
  mutate(
    ds     = as.Date(ds),
    y      = as.numeric(y),
    DperMar = readr::parse_number(as.character(DperMar))) %>%
  filter(!is.na(ds)) %>%
  arrange(ds) %>%
  as.data.frame()

# Holidays Dataset
holidays_df <- hol_raw %>%
  filter(country == "US") %>%     
  transmute(
    holiday = holiday,
    ds      = as.Date(ds, "%m/%d/%Y")
  )

# Model Parameters
m <- prophet::prophet(
  holidays = holidays_df,
  changepoint.range = .7,
  daily.seasonality = FALSE,
  weekly.seasonality = 4,
  yearly.seasonality = 10,
  changepoint.prior.scale = .05,
  seasonality.prior.scale = 10
)

# Define Regressors
m <- prophet::add_regressor(m, "DperMar", standardize = FALSE)

# Fitting
m <- prophet::fit.prophet(m, train)

# Forecasting
fc <- predict(m, R)

# Select Outputs
F <- fc %>% select(ds, yhat) %>% 
  rename(MAU = yhat)

# Excel Output
write.xlsx(list(
  Total_Daily=F,
  Total_Monthly=F %>% 
    mutate(Date=ceiling_date(ds,"month")-days(1)) %>% 
    group_by(Date) %>% 
    summarise(Total=sum(MAU),.groups="drop")), 
  paste0(out_dir,D_Today,"_Outputs_Total_Prophet_DperMar.xlsx"))

# Time Series Chart
print(
  plot(m, fc) +
    ggtitle(paste0("Forecast for Total ", metric)) +
    xlab("Year") +
    ylab("Total")
)

# Prophet Components Chart
print(
  prophet::prophet_plot_components(m, fc) +
    ggtitle(paste0("Prophet Components: ", metric))
)
